source/agora/common/MerkleTree.d

```diff
@@ -0,0 +1,215 @@
+/*******************************************************************************
+
+    Contains MerkleTree that makes binary tree of transactions.
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.node.MerkleTree;
+
+import agora.common.Data;
+import agora.common.Hash;
+import agora.common.Transaction;
+
+/// Ditto
+public struct MerkleTree
+{
+    /// The root of `MerkleTree`
+    public MerkleNode* root;
+
+    /// Array of `Transaction`
+    /// All transactions in the `MerkleTree` are stored here,
+    /// and refer to its memory everywhere.
+    public Transaction[] txs;
+
+    /// Array of MerkleNode,
+    /// All `MerkleNode` in the `MerkleTree` are stored here,
+    /// and refer to its memory everywhere.
+    public MerkleNode[] nodes;
```

Comment:
I don't see why you need a binary tree if you store everything in an array ? In fact I don't see why you would need the `MerkleNode` structure at all if you have an array of `Transaction` ?
Because as long as your tree is balanced (and it should be), you know how many hashes you are going to need.